nodeJS/express/controllers.md

```diff
@@ -0,0 +1,460 @@
+1. It then invokes a database query function `someDBQueryToGetUser` to retrieve the user data based on the userId.
+1. If the user is not found (`!user`), the controller sends a response with a 404 status code and the message `User not found`, using `res.status(404).send(...)`. It then returns from the controller function in order to not invoke any other logic in the controller, because sending a response does not stop the function execution itself.
+1. If the user is found, the controller sends a response with a 200 status code and the message `User found: ${user.name}`, using `res.status(200).send(...)`. This assumes that the user object has a name property.
+
+Very simple right?
+
+### Handling errors
+
+When building robust applications, it's crucial to handle errors gracefully within your application. By implementing proper error handling, you can provide meaningful error responses to the client and prevent your application from crashing unexpectedly. And as we know, errors may occur during async operations, and we are not exactly handling that in our previous code snippets. Let's see what we can do to handle these possibilities.
+
+#### try/catch
+
+Using the same code earlier, wrapping your controller logic in a `try/catch` is the quickest way to handle errors.
+
+```javascript
+// user controller file - controllers/userController.js
+
+const getUserById = async (req, res) => {
+  const userId = req.params.id;
+
+  try {
+    const user = await someDBQueryToGetUser(userId);
+
+    if (!user) {
+      res.status(404).send("User not found");
+      return;
+    }
+
+    res.send(`User found: ${user.name}`);
+  } catch (error) {
+    console.error("Error retrieving user:", error);
+    res.status(500).send("Internal Server Error");
+
+    // or we can call next(error) instead of sending a response here
+    // Using `next(error)` however will only render an error page in the express' default view and respond with the whole html to the client.
+    // So we will need to create a special type of middleware function if we want a different response and we will get to that in a bit.
+  }
+};
+```
+
+However, this is not exactly a clean solution since eventually, we will have to add the same try/catch block to *all* controllers. There is a package called [express-async-handler](https://www.npmjs.com/package/express-async-handler) that you can install to hide this bit. It essentially just chains a `catch` to "catch" any errors in the function.
+
+```javascript
+const asyncHandler = require("express-async-handler");
+
+// Any errors that is thrown in this function will automatically be caught and call the `next` function
+const getUserById = asyncHandler(async (req, res) => {
+  const userId = req.params.id;
+
+  const user = await someDBQueryToGetUser(userId);
+
+  if (!user) {
+    res.status(404).send("User not found");
+    return;
+  }
+
+  res.send(`User found: ${user.name}`);
+});
+```
+
```

Comment:
Can we add a lesson note of tip variant here? It should say something along the lines: 

> the asyncHandler function in the express-async-handler function is just 6 lines of code. try to take a guess on how it's implemented and then do [check out the source code](https://github.com/Abazhenov/express-async-handler/blob/master/index.js) for yourself.

Reply:
Should I remove `It essentially just chains a `catch` to "catch" any errors in the function` above and let them take a guess instead like described here?

Reply:
Removed it but let me know about your thoughts on that